R/expect-doppelganger.R

```diff
@@ -0,0 +1,209 @@
+#'   aspect of a plot, such as a slightly smaller or larger margin.
+#'   These changes will cause spurious failures because you need to
+#'   update your snapshots to reflect the upstream changes.
+#'
+#'   It would be distracting for both you and the CRAN maintainers if
+#'   such changes systematically caused failures on CRAN. This is why
+#'   snapshot expectations do not fail on CRAN by default and should
+#'   be treated as a monitoring tool that allows you to quickly check
+#'   how the appearance of your figures changes over time, and to
+#'   manually assess whether changes reflect actual problems in your
+#'   package.
+#'
+#'   Internally, this argument is passed to
+#'   [testthat::expect_snapshot_file()].
+#'
+#'
+#' @section Debugging:
+#' It is sometimes difficult to understand the cause of a failure.
+#' This usually indicates that the plot is not created
+#' deterministically. Potential culprits are:
+#'
+#' * Some of the plot components depend on random variation. Try
+#'   setting a seed.
+#'
+#' * The plot depends on some system library. For instance sf plots
+#'   depend on libraries like GEOS and GDAL. It might not be possible
+#'   to test these plots with vdiffr.
+#'
+#' To help you understand the causes of a failure, vdiffr
+#' automatically logs the SVG diff of all failures when run under R
+#' CMD check. The log is located in `tests/vdiffr.Rout.fail` and
+#' should be displayed on Travis.
+#'
+#' You can also set the `VDIFFR_LOG_PATH` environment variable with
+#' `Sys.setenv()` to unconditionally (also interactively) log failures
+#' in the file pointed by the variable.
+#'
+#' @examples
+#' if (FALSE) {  # Not run
+#'
+#' library("ggplot2")
+#'
+#' test_that("plots have known output", {
+#'   disp_hist_base <- function() hist(mtcars$disp)
+#'   expect_doppelganger("disp-histogram-base", disp_hist_base)
+#'
+#'   disp_hist_ggplot <- ggplot(mtcars, aes(disp)) + geom_histogram()
+#'   expect_doppelganger("disp-histogram-ggplot", disp_hist_ggplot)
+#' })
+#'
+#' }
+#' @export
+expect_doppelganger <- function(title,
+                                fig,
+                                path = deprecated(),
+                                ...,
+                                writer = write_svg,
+                                cran = FALSE) {
+  fig_name <- str_standardise(title)
+  testcase <- make_testcase_file(fig_name)
+  writer(fig, testcase, title)
+
+  if (!missing(...)) {
+    lifecycle::deprecate_soft(
+      "1.0.0",
+      "vdiffr::expect_doppelganger(... = )",
+    )
+  }
+  if (lifecycle::is_present(path)) {
+    lifecycle::deprecate_soft(
+      "1.0.0",
+      "vdiffr::expect_doppelganger(cran = )",
+    )
+  }
+
+  if (is_graphics_engine_stale()) {
+    testthat::skip(paste_line(
+      "The R graphics engine is too old.",
+      "Please update to R 4.1.0 and regenerate the vdiffr snapshots."
+    ))
+  }
+
+  testthat::local_edition(3)
```

Comment:
@hadley Is it correct to do this? This allows packages that haven't made the switch to testthat 3 to use vdiffr. For instance ggplot2 hasn't switched yet. This works well but maybe this breaks some contract with testthat?

Reply:
It seems reasonable to me.